zulip_bots/zulip_bots/bots/trivia_quiz/trivia_quiz.py

```diff
@@ -196,16 +198,40 @@ def format_quiz_for_markdown(quiz_id: str, quiz: Dict[str, Any]) -> str:
     )
     return content
 
-def grade_question(quiz: Dict[str, Any], answer: str) -> Tuple[bool, str]:
-    correct = (answer == quiz['correct_letter'])
+def update_quiz(quiz: Dict[str, Any], quiz_id: str, bot_handler: Any) -> None:
+    bot_handler.storage.put(quiz_id, json.dumps(quiz))
 
-    if correct:
-        long_answer = quiz['answers'][answer]
-        response = '**CORRECT!** {long_answer} :tada:'.format(long_answer=long_answer)
-        return correct, response
+def handle_answer(quiz: Dict[str, Any], option: str, quiz_id: str,
+                  bot_handler: Any) -> Tuple[bool, str]:
+    answer = quiz['answers'][quiz['correct_letter']]
+    start_new_question = False
+
+    if option not in quiz['answered_options'] and not quiz['ended']:
+        # The question has not been ended, so we store all the options
+        # which are being answered
+        quiz['answered_options'].append(option)
+        update_quiz(quiz, quiz_id, bot_handler)
+
+    if option == quiz['correct_letter']:
+        response = '**CORRECT!** {answer} :tada:'
+        # We don't start a new question if the question has already been
+        # answered
+        start_new_question = not quiz['ended']
+    else:
+        if len(quiz['answered_options']) == 3:
+            # We reveal the answer and start a new question upon having three failed
+            # attempts at answering a question
+            response = '''Three failed attempts at {id}:disappointed:
+The correct answer is {answer}.'''
+            start_new_question = True
+            # we reset this so that the users can keep answering a question after it has ended
```

Comment:
I think there's a simpler approach than clearing `answered_options` to allowing people to answer.  You don't really need to clear the data structure.  Instead, just accept answers forever, and separate your concerns:

* correct/disappointed -- just see if they answered the question right
* start new question? -- (answer_correct or >= 3 answers) and (still_active)
* update `answered_options` - just always do it unless answer has already been given

Reply:
Here's my suggestion of how to write it, and note that I make a little helper function that just takes care of message formatting (you'll have to flesh it out a bit, of course);

~~~ py
def build_response(is_correct, start_new_question, num_answers):
    if is_correct:
        response = '**CORRECT** ...'
    else:
        if num_answers >= 3:
            response = 'disappointed...correct answers is X'
        else:
            response = 'disappointed'

def handle_answer(...):
    is_new_answer = (option not in quiz['answered_options']
    if is_new_answer:
        quiz['answered_options'].append(new_answer)

    num_answers = len(quiz('answered_options'])

    is_correct = (option == quiz['correct_letter'])

    start_new_question = quiz['pending'] and (is_correct or num_answers >= 3)
    if start_new_question:
        quiz['pending'] = False

    if is_new_answer or start_new_question:
        update_quiz(quiz, quiz_id, bot_handler)

    response = build_response(is_correct, start_new_question):

    return start_new_question, response

Reply:
Right, I was clearing `answered_options` up everytime because I wanted two different error messages. I wanted the three failed attempt warning to appear only once, and all attempts after that all the warnings were same as before. But here, since we are changing the warning and just revealing the correct answer, so it would work.